app/src/main/kotlin/com/wire/android/ui/home/conversations/model/MessageTypes.kt

```diff
@@ -80,29 +87,90 @@ internal fun MessageBody(messageBody: MessageBody, onLongClick: (() -> Unit)? =
 fun MessageImage(
     rawImgData: ByteArray?,
     imgParams: ImageMessageParams,
+    assetUploadStatus: Message.UploadStatus,
     onImageClick: Clickable,
 ) {
+    val imageData: Bitmap? = if (rawImgData != null && rawImgData.size < MessageComposerViewModel.IMAGE_SIZE_LIMIT_BYTES)
+        rawImgData.toBitmap() else null
     Box(
         Modifier
             .clip(shape = RoundedCornerShape(dimensions().messageAssetBorderRadius))
+            .background(
+                color = MaterialTheme.wireColorScheme.onPrimary,
+                shape = RoundedCornerShape(dimensions().messageAssetBorderRadius)
+            )
+            .border(
+                width = 1.dp,
+                color = MaterialTheme.wireColorScheme.secondaryButtonDisabledOutline,
+                shape = RoundedCornerShape(dimensions().messageAssetBorderRadius)
+            )
+            .wrapContentSize()
             .combinedClickable(
                 enabled = onImageClick.enabled,
                 onClick = onImageClick.onClick,
                 onLongClick = onImageClick.onLongClick,
             )
     ) {
-        val imageData: Bitmap? =
-            if (rawImgData != null && rawImgData.size < MessageComposerViewModel.IMAGE_SIZE_LIMIT_BYTES) rawImgData.toBitmap() else null
+        when (assetUploadStatus) {
+            // Default states, we try to draw the image
+            Message.UploadStatus.NOT_UPLOADED -> {}
+            Message.UploadStatus.UPLOADED -> {
+                Image(
+                    painter = rememberAsyncImagePainter(imageData ?: getUriFromDrawable(LocalContext.current, R.drawable.ic_gallery)),
+                    contentDescription = stringResource(R.string.content_description_image_message),
+                    modifier = Modifier
+                        .align(Alignment.Center)
+                        .width(if (imageData != null) imgParams.normalizedWidth else dimensions().spacing24x)
+                        .height(if (imageData != null) imgParams.normalizedHeight else dimensions().spacing24x),
+                    alignment = Alignment.CenterStart,
+                    contentScale = ContentScale.Crop
+                )
+            }
+            // Trying to upload the asset
+            Message.UploadStatus.UPLOAD_IN_PROGRESS -> {
+                Column(
+                    verticalArrangement = Arrangement.Center,
+                    horizontalAlignment = Alignment.CenterHorizontally,
+                    modifier = Modifier
+                        .align(Alignment.Center)
+                        .size(dimensions().spacing200x)
```

Comment:
Can we take the actual dimensions at least for images and videos?

Reply:
Videos will be treated as generic files, but yeah, I can use the normalised width and height for images already when uploading it.